common/Obelisk/OAuth/State/Internal.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+{-| Internals of "Obelisk.OAuth.State". Use if you want to provide your own
+   `OAuthState` generator, but make sure you are using truly random values!
+-}
+module Obelisk.OAuth.State.Internal where
+
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+import           Data.Aeson             (FromJSON, ToJSON)
+import qualified Data.ByteString.Base16 as Base16
+import           Data.Text              (Text)
+import qualified Data.Text.Encoding     as T
+import           GHC.Generics           (Generic)
+import           System.Entropy         (getEntropy, getHardwareEntropy)
+
+
+-- | Oauth state to prevent CSRF attacks.
+newtype OAuthState = OAuthState { unOAuthState :: Text }
+  deriving (Generic, Show, Read, Eq, Ord, ToJSON, FromJSON)
+
+-- | Make an `OAuthState` from some Text.
+--
+--   If you need a brand new `OAuthState` you should use `genOAuthState`. This
+--   function is useful for deserialization of server responses for example.
+unsafeMkOAuthState :: Text -> OAuthState
+unsafeMkOAuthState = OAuthState
+
+-- | Get a `Text` representation of `OAuthState`.
+--
+--   Useful for serialization/encoding purposes.
+oAuthStateAsText :: OAuthState -> Text
+oAuthStateAsText = unOAuthState
+
+-- | Generate some truly random `OAuthState`.
+--
+--   By truly random we mean unguessable by an attacker - e.g. not dependent
+--   on some value derived from the current time.
+genOAuthState :: MonadIO m => m OAuthState
+genOAuthState = liftIO $ do
+  let fastRandom nr = maybe (getEntropy nr) pure =<< getHardwareEntropy nr
```

Comment:
Why are we getting entropy this way?  Why not just `getEntropy`?

Reply:
I just wanted it as random as possible (that way was suggested in the entropy docs) - it should not really matter though.

Reply:
That way was suggested in the `entropy` docs for a specific use case, not as the default.  Not everyone trusts the hardware RNG - this was the topic of extensive debate in the Linux kernel developer community.  `getEntropy` will get cryptographic-quality entropy in the standard way for the given system (usually `/dev/urandom` on linux), so that's what we should generally do.  Crypto needs to be kept simple.

Reply:
Yeah I know that debate. Sure I will go with just `getEntropy`!